xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h

```diff
@@ -489,27 +554,50 @@ struct AttentionBackwardKernel {
     using ThreadblockShape =
         cutlass::gemm::GemmShape<kBlockSizeI, kBlockSizeJ, GemmType::ThreadK>;
     using WarpShape = cutlass::gemm::GemmShape<32, 32, GemmType::WarpK>;
-    using DefaultMma = typename cutlass::gemm::threadblock::DefaultMma<
+
+    using ElementC = output_t;
+    using ElementAccum = accum_t;
+
+    // no-op output op - epilogue just stores result to global memory
+    using BiasGradEpilogueOutputOp = typename cutlass::epilogue::thread::LinearCombination<
+        ElementC,
+        DefaultConfig::EpilogueOutputOp::kCount,
+        typename DefaultConfig::EpilogueOutputOp::ElementAccumulator,
+        typename DefaultConfig::EpilogueOutputOp::ElementCompute,
+        cutlass::epilogue::thread::ScaleType::Nothing>;
+
+    using DefaultGemm = typename cutlass::gemm::kernel::DefaultGemm<
         scalar_t, // ElementA
         cutlass::layout::RowMajor, // LayoutA
         kIsAligned ? DefaultConfig::kAlignmentA : GemmType::kMinimumAlignment,
         scalar_t, // ElementB
         cutlass::layout::ColumnMajor, // LayoutB
         kIsAligned ? DefaultConfig::kAlignmentB : GemmType::kMinimumAlignment,
-        accum_t, // ElementC
+        ElementC, // ElementC
         cutlass::layout::RowMajor, // LayoutC
+        ElementAccum, // ElementAccumulator
         typename GemmType::OpClass,
         ArchTag,
         ThreadblockShape,
         WarpShape,
         typename GemmType::InstructionShape,
-        DefaultConfig::kStages,
+        BiasGradEpilogueOutputOp, // EpilogueOutputOp
+        void, // ThreadblockSwizzle (not used)
+        // multiple preloads, dropout Zij tile, and 3 stages push us over shared
+        // memory capacity on A100. set a ceiling on number of stages to save shared
+        // memory if dropout is in use.
+        kPreloadMmas && kApplyDropout ?
+            cutlass::const_min(2, DefaultConfig::kStages) :
+            DefaultConfig::kStages,  // Stages
```

Comment:
from looking at it seems like it does mean it selects the pipelined implementation. here's some benchmarks to show the impact

```
[---------------------------- attention backward (attn_bias=<class 'NoneType'>) -----------------------------]
                                                                            |  optimized[flshattB]  |  vanilla
1 threads: ---------------------------------------------------------------------------------------------------
      b16 B=8, M=512, H=64, K=64, p=0.0, BiasT=NoneType, BiasGrad=False     |           1.1         |     2.9
      b16 B=8, M=512, H=64, K=64, p=0.3, BiasT=NoneType, BiasGrad=False     |           1.2         |     3.4
      b16 B=8, M=512, H=64, K=128, p=0.0, BiasT=NoneType, BiasGrad=False    |           2.6         |     3.9
      b16 B=8, M=512, H=64, K=128, p=0.3, BiasT=NoneType, BiasGrad=False    |           2.7         |     4.3
      b16 B=8, M=1024, H=64, K=64, p=0.0, BiasT=NoneType, BiasGrad=False    |           4.1         |    10.0
      b16 B=8, M=1024, H=64, K=64, p=0.3, BiasT=NoneType, BiasGrad=False    |           4.4         |    12.0
      b16 B=8, M=1024, H=64, K=128, p=0.0, BiasT=NoneType, BiasGrad=False   |           9.1         |    12.0
      b16 B=8, M=1024, H=64, K=128, p=0.3, BiasT=NoneType, BiasGrad=False   |           9.4         |    13.9
      b16 B=16, M=512, H=64, K=64, p=0.0, BiasT=NoneType, BiasGrad=False    |           2.2         |     5.7
      b16 B=16, M=512, H=64, K=64, p=0.3, BiasT=NoneType, BiasGrad=False    |           2.4         |     6.7
      b16 B=16, M=512, H=64, K=128, p=0.0, BiasT=NoneType, BiasGrad=False   |           5.1         |     7.7
      b16 B=16, M=512, H=64, K=128, p=0.3, BiasT=NoneType, BiasGrad=False   |           5.3         |     8.7
      b16 B=16, M=1024, H=64, K=64, p=0.0, BiasT=NoneType, BiasGrad=False   |           8.2         |    20.1
      b16 B=16, M=1024, H=64, K=64, p=0.3, BiasT=NoneType, BiasGrad=False   |           8.7         |    23.9
      b16 B=16, M=1024, H=64, K=128, p=0.0, BiasT=NoneType, BiasGrad=False  |          18.2         |    24.0
      b16 B=16, M=1024, H=64, K=128, p=0.3, BiasT=NoneType, BiasGrad=False  |          18.8         |    27.8

Times are in milliseconds (ms).

[------------------------- attention backward (attn_bias=<class 'torch.Tensor'>) --------------------------]
                                                                          |  optimized[cutlassB]  |  vanilla
1 threads: -------------------------------------------------------------------------------------------------
      b16 B=8, M=512, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=False     |           2.1         |     2.9
      b16 B=8, M=512, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=True      |           2.1         |     2.9
      b16 B=8, M=512, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=False     |           3.1         |     3.4
      b16 B=8, M=512, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=True      |           3.1         |     3.4
      b16 B=8, M=512, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=False    |           3.6         |     3.9
      b16 B=8, M=512, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=True     |           3.6         |     3.9
      b16 B=8, M=512, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=False    |           4.5         |     4.3
      b16 B=8, M=512, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=True     |           4.5         |     4.3
      b16 B=8, M=1024, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=False    |           7.6         |    10.0
      b16 B=8, M=1024, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=True     |           7.6         |    10.0
      b16 B=8, M=1024, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=False    |          11.7         |    11.9
      b16 B=8, M=1024, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=True     |          11.7         |    11.9
      b16 B=8, M=1024, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=False   |          13.1         |    12.0
      b16 B=8, M=1024, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=True    |          13.1         |    12.0
      b16 B=8, M=1024, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=False   |          16.4         |    13.9
      b16 B=8, M=1024, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=True    |          16.4         |    13.9
      b16 B=16, M=512, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=False    |           3.9         |     5.7
      b16 B=16, M=512, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=True     |           4.0         |     5.7
      b16 B=16, M=512, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=False    |           6.0         |     6.7
      b16 B=16, M=512, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=True     |           6.0         |     6.7
      b16 B=16, M=512, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=False   |           7.3         |     7.7
      b16 B=16, M=512, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=True    |           7.3         |     7.7
      b16 B=16, M=512, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=False   |           8.9         |     8.7
      b16 B=16, M=512, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=True    |           8.9         |     8.7
      b16 B=16, M=1024, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=False   |          14.8         |    20.1
      b16 B=16, M=1024, H=64, K=64, p=0.0, BiasT=Tensor, BiasGrad=True    |          14.7         |    20.0
      b16 B=16, M=1024, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=False   |          23.1         |    23.9
      b16 B=16, M=1024, H=64, K=64, p=0.3, BiasT=Tensor, BiasGrad=True    |          23.1         |    23.9
      b16 B=16, M=1024, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=False  |          25.9         |    24.0
      b16 B=16, M=1024, H=64, K=128, p=0.0, BiasT=Tensor, BiasGrad=True   |          25.9         |    24.0
      b16 B=16, M=1024, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=False  |          32.4         |    27.8
      b16 B=16, M=1024, H=64, K=128, p=0.3, BiasT=Tensor, BiasGrad=True   |          32.4         |    27.8
```

Reply:
just curious, what are the plans to cut shmem usage?

Reply:
Which GPU did you run these benchmarks on? It's ~20-50% slower depending on the cases - might be worth investigating later if you want.


> just curious, what are the plans to cut shmem usage?

We are loading Q,K,dO from global memory twice. We could reuse the shared-memory to avoid that
Some illustration of the BW pass:
(1) Q@K, dO@V, gradV matmuls
![image](https://user-images.githubusercontent.com/43445237/208881148-fc5c36f3-63bb-48ba-a1e8-81dd9cf14acd.png)
(2) gradQ, gradK
![image](https://user-images.githubusercontent.com/43445237/208881205-49f018e6-b518-4007-a326-26eefa0e021d.png)

Reply:
## A100 Shmem occupancy - current situation (max ~160kb)
![image](https://user-images.githubusercontent.com/43445237/208883948-a12fea2a-3eb0-4084-a0bb-9775e1eb00b3.png)

## A100 Shmem occupancy - after potential changes (max ~131kb)
(1) setting `kNumStages=4` so the entire block of matrices fit in shared-memory
(2) loading dO/K from same shared-memory location (rather than re-loading them from global memory)
Also I expect this to be faster

![image](https://user-images.githubusercontent.com/43445237/208884260-96e5b7d5-4163-46d4-8d64-5bbe6ca98129.png)

Reply:
Just saw your question in https://github.com/NVIDIA/cutlass/issues/744, I believe it's related to this :) Let's do that as part of a different PR if possible to make things easier to review.
Also as a heads-up for this PR, I won't be available next week but hopefully we can get this merged early january!

Reply:
yes it was, thanks for the hints there! yes definitely we can put it in another PR, just wanted to start experimenting.

and sounds good, thanks for the quick reviews so far, happy holidays!